openssl: add enableFIPS feature

andrewzah opened the pull request:
OpenSSL provides a [validated](https://openssl-library.org/post/2025-03-11-fips-140-3/) FIPS cryptographic module for 3.x, supporting both [FIPS 140-2](https://csrc.nist.gov/pubs/fips/140-2/upd2/final) and [FIPS 140-3](https://csrc.nist.gov/pubs/fips/140-3/final) - (Federal Information Processing Standards). These are essentially cryptographic requirements for U.S. Federal departments and agencies.

This commit adds the `enableFIPS` parameter to `openssl` - set to `false` by default, which runs the [openssl-fipsinstall program](https://docs.openssl.org/master/man1/openssl-fipsinstall/) to generate the fipsmodule.cnf config file, and modifies openssl.cnf to add the FIPS provider.

I'd like to discuss supporting this feature, and get feedback on the modifications I made.

Further Information:
- https://en.wikipedia.org/wiki/FIPS_140
- https://docs.openssl.org/master/man7/fips_module/

~~This would close #324013.~~

---

```
$ echo 'abcd' | ./result-bin/bin/openssl md5

Error setting digest
40678A3ACE7F0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:375:Global default library context, Algorithm (MD5 : 100), Properties ()
40678A3ACE7F0000:error:03000086:digital envelope routines:evp_md_init_internal:initialization error:crypto/evp/digest.c:271:

$ echo 'abcd' | ./result-bin/bin/openssl sha256
SHA2-256(stdin)= fc4b5fd6816f75a7c81fc8eaa9499d6a299bd803397166e8c4cf9280b801d62c
```
```
$ ./result-bin/bin/openssl list -providers

Providers:
  base
    name: OpenSSL Base Provider
    version: 3.5.2
    status: active
  fips
    name: OpenSSL FIPS Provider
    version: 3.5.2
    status: active
```

---

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

IncredibleLaser commented:
Thanks for the efforts; one thing to consider though is that while all versions come with the food FIPS provider, only a selected sample of OpenSSL versions are actually FIPS certified, with the latest being version 3.1.2. The latest entry about that is at https://openssl-library.org/post/2025-03-11-fips-140-3/, if you look at https://openssl-library.org/source/ there is a dedicated section for FIPS certified versions. So if you're building with just enableFIPS, you're not actually compliant. While I don't think this option should ensure all that automatically, I think a warning during evaluation would help in case the version isn't actually certified.

IncredibleLaser commented:
What the section means is that you can build other OpenSSL versions to be compatible with the FIPS certified ones, since IIRC the latter has some algorithms usually not included. How I read it, the paragraph is purely about compatibility, not about certification.

From my experience, when you're using FIPS validated algorithms, it's because there's a requirement somewhere that mandates either the use of validated algorithms (which I'd say applies to my field) or the use of certified cryptographic libraries. Using a non-certified version with the FIPS provider is probably fine in the first case as long as someone can validate that you're using these algorithms, while the second case might be required in other constellations, e.g. in US government contracts.

thillux commented:
As already stated above, not every version of OpenSSL is also FIPS validated. Please at least state this in comments, that users should double check on the OpenSSL and NIST website, when doing more than testing or non-approved usage with the FIPS provider.

andrewzah commented:
@tiagolobocastro that makes sense to me.

thillux reviewed (approved):
LGTM. Please rebase.

andrewzah committed:
openssl: add FIPS features, add openssl 3.1.2 drv

* add enableFIPS
* add a build of 3.1.2, which is the latest validated version for
  FIPS 140-3
* add a check for the openssl version that warns if FIPS is enabled for
  a non-validated version
* add a check if enableFIPS is enabled to make sure that:
  1. the FIPS provider is listed in `openssl list -providers`
  2. disallowed algorithms like MD5 fail, e.g. `echo '1' | openssl md5`
  3. openssl works normally, e.g. `echo '1' | openssl sha256`
* add a use-etc-ssl-certs patch for openssl 3.1.x

andrewzah committed:
openssl_fips: init at 3.1.2

* openssl 3.1.2 is the latest FIPS validated version

h0nIg commented on pkgs/development/libraries/openssl/3.1/use-etc-ssl-certs.patch:
```diff
@@ -0,0 +1,12 @@
+ #  define X509_CERT_AREA          OPENSSLDIR
+ #  define X509_CERT_DIR           OPENSSLDIR "/certs"
+-#  define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
++#  define X509_CERT_FILE          "/etc/ssl/certs/ca-certificates.crt"
```
```suggestion
+#  define X509_CERT_FILE          OPENSSLDIR "/cert.pem"
```

and drv:

```
postPatch = ''
  substituteInPlace include/internal/cryptlib.h \
    --replace-fail "OPENSSLDIR \"/cert.pem\"" "${pkgs.cacert}/etc/ssl/certs/ca-bundle.crt"
'';```

thillux replied:
Please don't disable the global CA certificates here. "/etc/ssl/certs/ca-certificates.crt" is correct.

andrewzah replied:
Ah, okay.

rhendric commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -277,6 +287,29 @@ let
       '';
 
+      doInstallCheck = enableFIPS;
+      installCheckPhase = lib.optionalString enableFIPS ''
```
This suggestion may not work, but in principle, this PR shouldn't cause everything to rebuild because it's only adding a feature flag that defaults to false, right? I think this line is the only reason that it does trigger rebuilds, and I think you could fix that with:
```suggestion
      ${if enableFIPS then "installCheckPhase" else null} = ''
```

ShamrockLee reviewed (changes requested):
Sorry for my delayed response.

> I'll remove the reference to openssl 3.1.2 in all-packages.nix. I'm also fine with marking openssl-fips (v3.1.2) as insecure.

[pkgs/README.md](https://github.com/NixOS/nixpkgs/tree/master/pkgs#quick-start-to-adding-a-package) explicitly advise against introducing new packages with known vulnerabilities.

Here are some possible ways that you could push this:
1. Maintain a properly patched OpenSSL 3.3.6 in Nixpkgs and keep its security patches up-to-date.
   - You will likely become a maintainer of the poenssl 3.3.6 package responsible for updating the patches.
   - Ontroducing a new openssl version requires areement from current maintainers and the @NixOS/security-review team.
2. Maintain them out-of-tree (probably inside a flake under your control), and optionally implement mechanisms on the Nixpkgs side to facillitate such down-stream usage.
   - We coul discuss about what to include.
     E.g., implement the `enableFIPS` logic here and the actual package in the downstream flake.
3. Implement the `enableFIPS` logic here and expect the operators to construct working packages.
   - The change would bw challenging to test and require sophisticated test logic maintainers can run easily.

andrewzah commented:
@ShamrockLee Looking at this PR again, it seems better to keep the fips logic in a separate public flake.

We could keep `enableFIPS`, but it'd add a lot of logic and testing for something that *very* few people would use, outside of a specific government context. I guess it depends on how big of a scope the security-review team has in mind for openssl.

ShamrockLee commented:
I don't have the required background knowledge to assess it, but what @h0nIg mentioned seem like a wonderful solution that allows us to implement `enableFIPS` on newer OpenSSL versions without the security overhead of the whole outdated OpenSSL version.

@andrewzah Is that the case?

andrewzah commented:
> If you require FIPS-approved functionality, it is essential to build your FIPS provider using one of the validated versions listed there. Normally, it is possible to utilize a FIPS provider constructed from one of the validated versions alongside libcrypto and libssl compiled from any release within the same **major release series**. This flexibility enables you to address bug fixes and CVEs that fall outside the FIPS boundary.

Yes, If I understand this correctly, a fips provider built using 3.1.2 can be used by other versions in the same major release: e.g. `3.4.6`, `3.5.7` (LTS), and `3.6.3`. But not for `4.0.1`, which released on June 9th this year.

The [openssl download page](https://openssl-library.org/source/) makes it more clear:

> Other OpenSSL Releases MAY use the validated FIPS provider, but **MUST NOT build and use their own FIPS provider**. For example you can build OpenSSL 3.4 and use the OpenSSL 3.0.9 FIPS provider with it.

So we could make a 3.1.2-built fips module available via `enableFIPS`, but only for 3.x versions.

@ShamrockLee @h0nIg

ShamrockLee commented:
That looks promising!
